src/main/scala/uk/gov/nationalarchives/omega/api/services/ApiService.scala

```diff
@@ -0,0 +1,146 @@
+import uk.gov.nationalarchives.omega.api.conf.ServiceConfig
+import uk.gov.nationalarchives.omega.api.connectors.JmsConnector
+import uk.gov.nationalarchives.omega.api.services.LocalMessage.createLocalMessage
+import uk.gov.nationalarchives.omega.api.services.ServiceState.{ Started, Starting, Stopped, Stopping }
+
+import java.nio.file.{ Files, Paths }
+import java.util.UUID
+
+class ApiService(val config: ServiceConfig) extends Stateful {
+
+  def start: IO[ExitCode] =
+    // check if we can start the service
+    switchState(Stopped, Starting).ifM(attemptStartUp(), IO.pure(ExitCode(invalidState)))
+
+  private def attemptStartUp(): IO[ExitCode] =
+    // switch to the started state
+    switchState(Starting, Started)
+      .ifM(
+        doStart(),
+        IO.pure(ExitCode(invalidState))
+      )
+      .handleErrorWith(error =>
+        logger.error(s"Shutting down due to ${error.getMessage}") *>
+          doStop() *> switchState(Starting, Stopped) *> IO.pure(ExitCode.Error)
+      )
+
+  def stop(): IO[ExitCode] =
+    // check if we can stop the service
+    switchState(Started, Stopping).ifM(
+      logger.info("Closing connection..") *>
+        switchState(Stopping, Stopped).ifM(doStop(), IO.pure(ExitCode(invalidState))),
+      IO.pure(ExitCode(invalidState))
+    )
+
+  private def doStart(): IO[ExitCode] = {
+
+    // TODO(AR) - one client, how to ack a consumer message after local persistence and then process it, and then produce a response
+    // TODO(AR) how to wire up queues and services using a config file or DSL?
+    // TODO(AR) request queue will typically be 1 (plus maybe a few more for expedited ops), response queues will be per external application
+    val localQueue: IO[Queue[IO, LocalMessage]] = Queue.bounded[IO, LocalMessage](config.maxLocalQueueSize)
+    val jmsConnector = new JmsConnector(config)
+    getLocalMessageStore.flatMap {
+      case Right(localMessageStore) =>
+        val result = for {
+          q <- localQueue
+          res <-
+            jmsConnector.getJmsProducerAndConsumer(QueueName(config.requestQueue)).use {
+              case (jmsProducer, jmsConsumer) =>
+                process(q, jmsConsumer, jmsProducer, localMessageStore)
+            }
+        } yield res
+        result.as(ExitCode.Success)
+      case Left(exitCode) => IO.pure(exitCode)
+    }
+  }
+
+  private def getLocalMessageStore: IO[Either[ExitCode, LocalMessageStore]] = IO {
+    val messageStoreFolder = Paths.get(config.tempMessageDir)
+    Files.createDirectories(messageStoreFolder)
+    Right(new LocalMessageStore(messageStoreFolder))
+  }.handleErrorWith(ex =>
+    logger.error(s"Failed to created local message store due to ${ex.getMessage}") *> IO.pure(Left(ExitCode.Error))
+  )
+
+  private def process(
+    queue: Queue[IO, LocalMessage],
+    consumer: JmsAcknowledgerConsumer[IO],
+    producer: JmsProducer[IO],
+    localMessageStore: LocalMessageStore
+  ) =
+    IO.race(
```

Comment:
I don't think `IO.race` is what we want here. From the docs:
> Run two IO tasks concurrently, and return the first to finish, either in success or error. The loser of the race is canceled.

We always want both of these task to complete, I imagine. I'm not sure how this is working right now. Are there tests for `ApiService`?

Reply:
The common component between the message handler and the dispatcher is the queue. The handler puts messages onto the queue and the dispatcher takes them off. We do indeed want these to run forever, the only reason that one would finish is in error, in which case the other is cancelled.

There are no tests currently for `ApiService` which is due to the difficulty in creating or mocking instances of `JmsMessage` which has a private constructor within the jms4s library.

Reply:
I mean that `IO.race` runs two tasks concurrently, and then when either task finishes, the other one in cancelled. Is this what we want?